File: src/utils/dateUtils.ts
import {
  differenceInCalendarDays,
  addWeeks,
  addMonths,
  addYears,
  addDays,
  parseISO,
  format,
} from 'date-fns';
import type { Subscription } from '../types';

export function getDaysUntilRenewal(nextRenewalDate: string): number {
  return differenceInCalendarDays(parseISO(nextRenewalDate), new Date());
}

export function advanceByOneCycle(sub: Subscription): string {
  const base = parseISO(sub.nextRenewalDate);
  let next: Date;

  switch (sub.billingCycle) {
    case 'weekly':
      next = addWeeks(base, 1);
      break;
    case 'monthly':
      next = addMonths(base, 1);
      break;
    case 'yearly':
      next = addYears(base, 1);
      break;
    case 'custom': {
      const amount = sub.customCycleAmount ?? 1;
      const unit = sub.customCycleUnit ?? 'months';
      if (unit === 'days') next = addDays(base, amount);
      else if (unit === 'weeks') next = addWeeks(base, amount);
      else next = addMonths(base, amount);
      break;
    }
  }

  return format(next, 'yyyy-MM-dd');
}

export function formatRenewalDate(dateStr: string): string {
  return format(parseISO(dateStr), 'MMM d, yyyy');
}

export function formatDaysUntil(days: number): string {
  if (days < 0) return `${Math.abs(days)} day${Math.abs(days) !== 1 ? 's' : ''} overdue`;
  if (days === 0) return 'Today';
  if (days === 1) return 'Tomorrow';
  return `In ${days} days`;
}

export function getUrgency(days: number): 'overdue' | 'critical' | 'warning' | 'upcoming' | 'ok' {
  if (days < 0) return 'overdue';
  if (days === 0) return 'critical';
  if (days <= 3) return 'critical';
  if (days <= 7) return 'warning';
  if (days <= 14) return 'upcoming';
  return 'ok';
}

File: android-app/app/src/main/kotlin/com/subhandler/utils/DateUtils.kt
package com.subhandler.utils

import com.subhandler.model.BillingCycle
import com.subhandler.model.CustomCycleUnit
import com.subhandler.model.Subscription
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit

private val DATE_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd")
private val DISPLAY_FORMAT = DateTimeFormatter.ofPattern("MMM d, yyyy")

fun parseDate(dateStr: String): LocalDate = LocalDate.parse(dateStr, DATE_FORMAT)

fun formatDate(dateStr: String): String =
    parseDate(dateStr).format(DISPLAY_FORMAT)

fun todayString(): String = LocalDate.now().format(DATE_FORMAT)

fun getDaysUntil(nextRenewalDate: String): Long {
    val today = LocalDate.now()
    val renewal = parseDate(nextRenewalDate)
    return ChronoUnit.DAYS.between(today, renewal)
}

fun formatDaysUntil(days: Long): String = when {
    days < 0  -> "${-days} day${if (-days == 1L) "" else "s"} overdue"
    days == 0L -> "Today"
    days == 1L -> "Tomorrow"
    else       -> "In $days days"
}

enum class Urgency { OVERDUE, CRITICAL, WARNING, UPCOMING, OK }

fun getUrgency(days: Long): Urgency = when {
    days < 0  -> Urgency.OVERDUE
    days == 0L -> Urgency.CRITICAL
    days <= 3 -> Urgency.CRITICAL
    days <= 7 -> Urgency.WARNING
    days <= 14 -> Urgency.UPCOMING
    else       -> Urgency.OK
}

fun advanceByOneCycle(subscription: Subscription): String {
    val date = parseDate(subscription.nextRenewalDate)
    val newDate = when (subscription.billingCycle) {
        BillingCycle.WEEKLY  -> date.plusWeeks(1)
        BillingCycle.MONTHLY -> date.plusMonths(1)
        BillingCycle.YEARLY  -> date.plusYears(1)
        BillingCycle.CUSTOM  -> {
            val amount = subscription.customCycleAmount?.toLong() ?: 1L
            when (subscription.customCycleUnit ?: CustomCycleUnit.MONTHS) {
                CustomCycleUnit.DAYS   -> date.plusDays(amount)
                CustomCycleUnit.WEEKS  -> date.plusWeeks(amount)
                CustomCycleUnit.MONTHS -> date.plusMonths(amount)
            }
        }
    }
    return newDate.format(DATE_FORMAT)
}

/** Returns triggerDate (yyyy-MM-dd) = nextRenewalDate - reminderDays */
fun reminderDate(subscription: Subscription): LocalDate? {
    if (subscription.reminderDays <= 0) return null
    return parseDate(subscription.nextRenewalDate).minusDays(subscription.reminderDays.toLong())
}

fun formatCycleLabel(subscription: Subscription): String = when (subscription.billingCycle) {
    BillingCycle.WEEKLY  -> "/wk"
    BillingCycle.MONTHLY -> "/mo"
    BillingCycle.YEARLY  -> "/yr"
    BillingCycle.CUSTOM  -> {
        val amount = subscription.customCycleAmount ?: 1
        val unit = subscription.customCycleUnit?.label ?: "months"
        "/$amount $unit"
    }
}
